`postinstall` in `package.json` scripts seems not to be executed anymore in VAADIN 23+ with `vaadin-maven-plugin:build-frontend`

### Description of the bug

This is a bug regarding the VAADIN Maven Plugin. I hope this is the right place, I couldn't find a dedicated GitHub repository for it.

In our project, we adjusted the `tsconfig.json` to use a custom output directory for compiled JS files. This was set to `frontend/target` and worked for years now including VAADIN 22.0.12. After upgrading to VAADIN 23, this folder won't be created anymore.

It was created when building the application in production mode / starting the application in non-production mode.

We rely on a JS file in the `frontend/target` folder to be present with using `@JsModule("./target/folder1/component.js")` in a Java class (see Minimal reproducible example).

As said, since VAADIN 23, this folder won't be created anymore. Due to that, the required JS file cannot be found leading to a failure of the `build-frontend` goal of the VAADIN Maven Plugin.

### Error Log

```log
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:23.0.5:build-frontend (default) on project %project name%: Could not execute build-frontend goal: Error occured during goal execution:
[ERROR]
[ERROR]   Failed to find the following files:
[ERROR]       - ./target/folder1/component.js
[ERROR]
[ERROR]   Locations searched were:
[ERROR]       - `C:\Users\%username%\[...]\project\frontend` in this project
[ERROR]       - `META-INF/frontend` in included JARs
[ERROR]       - `META-INF/resources/frontend` in included JARs
```

### Debugging

From debugging I noticed that **the `postinstall` defined in the `package.json` scripts isn't executed anymore since VAADIN 23**. There we have `"postinstall": "tsc"` which should create the `frontend/target` folder which is missing now (see Minimal reproducible example). This is most likely the issue for the bug.

**VAADIN 22:**
```log
[INFO] Running `npm install` to resolve and optionally download frontend dependencies. This may take a moment, please stand by...
[DEBUG] 
C:\Program Files\nodejs\npm.cmd --no-update-notifier \
    --no-audit --scripts-prepend-node-path=true install

[DEBUG] Output of `C:\Program Files\nodejs\npm.cmd --no-update-notifier --no-audit --scripts-prepend-node-path=true install`:
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm WARN deprecated core-js@2.6.12: core-js@<3.4 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up 
to 100x even if nothing is polyfilled. Please, upgrade your dependencies to the actual version of core-js.
[DEBUG] 
[DEBUG] > postinstall
[DEBUG] > tsc
[DEBUG]
[DEBUG] 
[DEBUG] added 1436 packages in 3m
[DEBUG]
[DEBUG] 102 packages are looking for funding
[DEBUG]   run `npm fund` for details
[INFO] Frontend dependencies resolved successfully.
```

**VAADIN 23:**
```log
[INFO] Running `npm install` to resolve and optionally download frontend dependencies. This may take a moment, please stand by...
[INFO] Globally node version 16.13.1 is older than 16.14.0. Using node from C:\Users\%username%\.vaadin.
[DEBUG] Using node 16.14.0 located at C:\Users\%username%\.vaadin\node\node.exe
[DEBUG] Using npm 8.3.1 located at C:\Users\%username%\.vaadin\node\node.exe
[DEBUG] 
C:\Users\%username%\.vaadin\node\node.exe \
    C:\Users\%username%\.vaadin\node\node_modules\npm\bin\npm-cli.js \
    --no-update-notifier --no-audit \
    --scripts-prepend-node-path=true --ignore-scripts \
    install

[INFO] using 'C:\Users\%username%\.vaadin\node\node.exe C:\Users\%username%\.vaadin\node\node_modules\npm\bin\npm-cli.js --no-update-notifier --no-audit --scripts-prepend-node-path=true --ignore-scripts install' for frontend package installation
[WARNING] package-lock.json is missing from this project. This may cause the npm package installation to take several minutes. It is recommended to keep the package-lock.json file persistently in your project. Please stand by...    
[DEBUG] Output of `C:\Users\%username%\.vaadin\node\node.exe C:\Users\%username%\.vaadin\node\node_modules\npm\bin\npm-cli.js --no-update-notifier --no-audit --scripts-prepend-node-path=true --ignore-scripts install`:
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm WARN deprecated core-js@2.6.12: core-js@<3.4 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up 
to 100x even if nothing is polyfilled. Please, upgrade your dependencies to the actual version of core-js.
[DEBUG] 
[DEBUG] added 1208 packages in 3m
[DEBUG]
[DEBUG] 101 packages are looking for funding
[DEBUG]   run `npm fund` for details
[INFO] Frontend dependencies resolved successfully.
```


### Expected behavior

The folder should be created by `tsc` after building in production mode or starting the application in non-production mode.

### Minimal reproducible example

I made an example project showcasing it: [missing-target-folder.zip](https://github.com/vaadin/flow/files/8482105/missing-target-folder.zip)


Note: I downloaded the project from https://vaadin.com/docs/latest/flow/guide/quick-start and adjusted it.

Simply run `mvn clean install -Pproduction` to build it.

It has a git repository so you can have a look at `229bdcaf2dfb9206ef4fa1021787cd5a738281a0` to see what I exactly changed. This commit uses VAADIN 22.0.12. The build will succeed on that commit.

In commit `95f7d3e33b7eeb67f802d7aa6e86f13632d402aa`, I set the version back to VAADIN 23.0.5. The build will fail on that commit.

Important parts:
- Added a dummy TS and Java file.
- Added our `scripts` section in the `package.json`.
- Adjustments to the `tsconfig.json`.
- Set `pnpm.enable` to `false` in the `pom.xml`.

### Versions

- Vaadin / Flow version: 23.0.0+
- Java version: 17.0.2 2022-01-18 LTS
- OS version: Windows 10 21H1
- Browser version (if applicable): Not needed.
- Application Server (if applicable): Not needed.
- IDE (if applicable): Not needed.

I already tried that but it wasn't executed at all, so I thought it was the wrong place.

```log
// same as above
[DEBUG] added 1304 packages in 22s
[DEBUG]
[DEBUG] 101 packages are looking for funding
[DEBUG]   run `npm fund` for details
[INFO] Frontend dependencies resolved successfully.
[DEBUG] Running postinstall for 'esbuild'
[DEBUG] Running postinstall for '@vaadin/vaadin-usage-statistics'
[INFO] Copying frontend resources from jar files ...
[INFO] Visited 220 resources. Took 100 ms.
[DEBUG] Found no local frontend resources for the project
[DEBUG] Updated C:\Users\%username%\[...]\project\target\frontend\generated-flow-imports-fallback.js
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  58.562 s
[INFO] Finished at: 2022-04-14T11:46:15+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:23.0.4:build-frontend (default) on project daem: Could not execute build-frontend goal: Error oc
cured during goal execution:
[ERROR]
[ERROR]   Failed to find the following files:
[ERROR]       - ./target/folder1/component.js
[ERROR]
[ERROR]   Locations searched were:
[ERROR]       - `C:\Users\%username%\[...]\project\frontend` in this project
[ERROR]       - `META-INF/frontend` in included JARs
[ERROR]       - `META-INF/resources/frontend` in included JARs
```

Do I have to do something in addition to that like building it differently?

I already tried that but it wasn't executed at all, so I thought it was the wrong place.

```log
// same as above
[DEBUG] added 1304 packages in 22s
[DEBUG]
[DEBUG] 101 packages are looking for funding
[DEBUG]   run `npm fund` for details
[INFO] Frontend dependencies resolved successfully.
[DEBUG] Running postinstall for 'esbuild'
[DEBUG] Running postinstall for '@vaadin/vaadin-usage-statistics'
[INFO] Copying frontend resources from jar files ...
[INFO] Visited 220 resources. Took 100 ms.
[DEBUG] Found no local frontend resources for the project
[DEBUG] Updated C:\Users\%username%\[...]\project\target\frontend\generated-flow-imports-fallback.js
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  58.562 s
[INFO] Finished at: 2022-04-14T11:46:15+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:23.0.4:build-frontend (default) on project %project name%: Could not execute build-frontend goal: Error occured during goal execution:
[ERROR]
[ERROR]   Failed to find the following files:
[ERROR]       - ./target/folder1/component.js
[ERROR]
[ERROR]   Locations searched were:
[ERROR]       - `C:\Users\%username%\[...]\project\frontend` in this project
[ERROR]       - `META-INF/frontend` in included JARs
[ERROR]       - `META-INF/resources/frontend` in included JARs
```

Do I have to do something in addition to that like building it differently?

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
@@ -415,26 +415,21 @@ public class TaskRunNpmInstall implements FallibleCommand {
         }
 
         List<String> postinstallPackages = new ArrayList<>();
+        postinstallPackages.add(".");
         postinstallPackages.add("esbuild");
         postinstallPackages.add("@vaadin/vaadin-usage-statistics");
         postinstallPackages.addAll(additionalPostinstallPackages);
 
         for (String postinstallPackage : postinstallPackages) {
-            if (postinstallPackage.trim().equals("")) {
+            File packageJsonFile = getPackageJsonForModule(postinstallPackage);
+            if (packageJsonFile == null || !packageJsonFile.exists()) {
                 continue;
             }
+            File packageFolder = packageJsonFile.getParentFile();
 
-            // Execute "npm run postinstall" in the desired folders in
-            // node_modules
-            File packageFolder = new File(packageUpdater.nodeModulesFolder,
-                    postinstallPackage);
-            if (!packageFolder.exists()) {
-                continue;
-            }
             try {
                 JsonObject packageJson = TaskGeneratePackageJson
-                        .getJsonFileContent(
-                                new File(packageFolder, "package.json"));
+                        .getJsonFileContent(packageJsonFile);
                 if (!containsPostinstallScript(packageJson)) {
                     logger.debug(
                             "Skipping postinstall for '{}' as no postinstall script was found in the package.json",
@@ -465,6 +460,20 @@ public class TaskRunNpmInstall implements FallibleCommand {
         }
         lastInstallStats.installTimeMs = System.currentTimeMillis() - startTime;
         lastInstallStats.packageManager = enablePnpm ? "pnpm" : "npm";
+
+    }
+
+    private File getPackageJsonForModule(String module) {
+        if (module.trim().equals("")) {
+            return null;
+        }
+        if (module.equals(".")) {
+            // The location of the project package.json
+            return new File(packageUpdater.npmFolder, "package.json");
+        }
+
+        return new File(new File(packageUpdater.nodeModulesFolder, module),
+                "package.json");
 
     }
 

--- a/flow-tests/test-frontend/vite-basics/package.json
+++ b/flow-tests/test-frontend/vite-basics/package.json
@@ -1,6 +1,9 @@
 {
   "name": "no-name",
   "license": "UNLICENSED",
+  "scripts": {
+    "postinstall": "node -e \"fs.writeFileSync('target/classes/main.postinstall', 'hello')\""
+  },
   "dependencies": {
     "@polymer/polymer": "3.4.1",
     "@testscope/all": "../vite-test-assets/packages/@testscope/all",

--- a/flow-tests/test-frontend/vite-basics/src/test/java/com/vaadin/viteapp/PostinstallIT.java
+++ b/flow-tests/test-frontend/vite-basics/src/test/java/com/vaadin/viteapp/PostinstallIT.java
@@ -1,0 +1,19 @@
+package com.vaadin.viteapp;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class PostinstallIT extends ViteDevModeIT {
+
+    @Test
+    public void postinstallRanForProject() throws IOException {
+        waitForDevServer(); // This is what runs the postinstall script
+        Assert.assertEquals("hello", IOUtils.toString(
+                getClass().getClassLoader().getResource("main.postinstall"),
+                StandardCharsets.UTF_8));
+    }
+}

